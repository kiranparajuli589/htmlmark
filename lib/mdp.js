import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
