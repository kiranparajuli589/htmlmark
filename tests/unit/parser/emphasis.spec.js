import { MDP } from "../../../lib/mdp.js"

const mdp = new MDP()


describe("Emphasis Parsing", () => {
	describe("bold", () => {
		it("should match the consecutive occurrences", () => {
			const lines = [
				"**first bold** **second bold** **third bold**"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should be deep tokenized", () => {
			const lines = [
				"**bold with `code` and *italics* and ~~strikes~~ and [link](https://kiranparajuli.com.np)**"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
	})
	describe("italics", () => {
		it("should match the consecutive occurrences", () => {
			const lines = [
				"*first italics* *second italics* *third italics*"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should be deep tokenized", () => {
			const lines = [
				"*italics with `code` and ~~strikes~~ and [link](https://kiranparajuli.com.np)*"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
	})
	describe("strikes", () => {
		it("should match the consecutive occurrences", () => {
			const lines = [
				"~~first strikes~~ ~~second strikes~~ ~~third strikes~~"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should be deep tokenized", () => {
			const lines = [
				"~~strikes with ~ gem and `code` and **bold** and *italics* and [link](https://kiranparajuli.com.np)~~"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
	})
	describe("code", () => {
		it("should match the consecutive occurrences", () => {
			const lines = [
				"`first code` `second code` `third code`"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should not be deep tokenized", () => {
			const lines = [
				"`code with ~ gem and **bold** and *italics* and [link](https://kiranparajuli.com.np)`"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should not panic with angles inside", () => {
			const lines = [
				"Pre-formatted code blocks are used for writing about programming or " +
				"markup source code. Rather than forming normal paragraphs, the lines " +
				"of a code block are interpreted literally. Markdown wraps a code block " +
				"in both `<pre>` and `<code>` tags."
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
	})
	describe("link", () => {
		it("should match the consecutive occurrences", () => {
			const lines = [
				"[link](https://kiranparajuli.com.np) [link](https://kiranparajuli.com.np)"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
		it("should deep tokenized the link title", () => {
			const lines = [
				"[*italic* with **bold** and `code`](https://kiranparajuli.com.np)"
			]
			const lexerData = mdp.h(lines)
			expect(lexerData).toMatchSnapshot()
		})
	})
})
