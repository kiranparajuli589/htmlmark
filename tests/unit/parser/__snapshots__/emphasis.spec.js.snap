// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emphasis Parsing bold should be deep tokenized 1`] = `"<p><strong>bold with <code>code</code> and <em>italics</em> and <s>strikes</s> and <a href=\\"https://kiranparajuli.com.np\\">link</a></strong></p>"`;

exports[`Emphasis Parsing bold should match the consecutive occurrences 1`] = `"<p><strong>first bold</strong> <strong>second bold</strong> <strong>third bold</strong></p>"`;

exports[`Emphasis Parsing code should match the consecutive occurrences 1`] = `"<p><code>first code</code> <code>second code</code> <code>third code</code></p>"`;

exports[`Emphasis Parsing code should not be deep tokenized 1`] = `"<p><code>code with ~ gem and **bold** and *italics* and [link](https://kiranparajuli.com.np)</code></p>"`;

exports[`Emphasis Parsing code should not panic with angles inside 1`] = `"<p>Pre-formatted code blocks are used for writing about programming or markup source code. Rather than forming normal paragraphs, the lines of a code block are interpreted literally. Markdown wraps a code block in both <code>&lt;pre&gt;</code> and <code>&lt;code&gt;</code> tags.</p>"`;

exports[`Emphasis Parsing italics should be deep tokenized 1`] = `"<p><em>italics with <code>code</code> and <s>strikes</s> and <a href=\\"https://kiranparajuli.com.np\\">link</a></em></p>"`;

exports[`Emphasis Parsing italics should match the consecutive occurrences 1`] = `"<p><em>first italics</em> <em>second italics</em> <em>third italics</em></p>"`;

exports[`Emphasis Parsing link should deep tokenized the link title 1`] = `"<p><a href=\\"https://kiranparajuli.com.np\\"><em>italic</em> with <strong>bold</strong> and <code>code</code></a></p>"`;

exports[`Emphasis Parsing link should match the consecutive occurrences 1`] = `"<p><a href=\\"https://kiranparajuli.com.np\\">link</a> <a href=\\"https://kiranparajuli.com.np\\">link</a></p>"`;

exports[`Emphasis Parsing strikes should be deep tokenized 1`] = `"<p><s>strikes with ~ gem and <code>code</code> and <strong>bold</strong> and <em>italics</em> and <a href=\\"https://kiranparajuli.com.np\\">link</a></s></p>"`;

exports[`Emphasis Parsing strikes should match the consecutive occurrences 1`] = `"<p><s>first strikes</s> <s>second strikes</s> <s>third strikes</s></p>"`;
